### PR TITLE
Add ICEUpdater Output for Real-time MOOSE updates to ICE.

### PR DESCRIPTION
**Pull request by jwpeterson**
Remaining issues:
- [x] The noproxy thing should be made into a runtime option.  
- [x] Restrict the ice updater test in the test harness to only run when `LIBMESH_HAVE_CURL` is defined.  
- [x] When the ice updater test passes, it prints a bunch of output to the screen that we don't want -- the default for the test harness is to suppress all output for successful runs. 
- [x] Can't rely on having TBB, need to guard with `LIBMESH_HAVE_TBB_API` or use abstractions.

---

Additional notes:

Cherry-picked @amccaskey's relevant commits from #5716 that had an insane history.  This PR replaces PR #5716 and addresses #5663.

For the record, the steps taken were:

```
git cherry-pick 288a429 # fix conflict in Moose.C
git cherry-pick 6e45108 # remove the signed-off-by
git cherry-pick 93eb354 # remove the signed-off-by
# skip next three (repeated) commits and 2 merges
git cherry-pick a53b26d # fix conflict in Moose.C
git cherry-pick fdb8805 # fix conflict in ICEUpdaterTester.py
git cherry-pick 24eb943 # fix conflict in ICEUpdater.h
git cherry-pick a9ea92f # fix conflict in ICEUpdaterTester.py
git cherry-pick d522a94
git cherry-pick 5f0eb61 # drop incorrect change to Moose.C
git cherry-pick b524abc
git cherry-pick b465a3e # empty
git cherry-pick 1ebf4dc # fix conflicts in ICEUpdater.C
```

@amccaskey note that I get a Python error (Python actually crashes) while running the new iceupdater test (not sure if it's due to cherry-picking or actual errors in the PR):

```
outputs/iceupdater.testupdater.................................. FAILED (DID NOT GET CORRECT NUMBER OF POSTS)
Traceback (most recent call last):
  File "./run_tests", line 24, in <module>
    TestHarness.buildAndRun(sys.argv, app_name, MOOSE_DIR)
  File "/Users/petejw/projects/moose_git/python/TestHarness/TestHarness.py", line 34, in buildAndRun
    harness.findAndRunTests()
  File "/Users/petejw/projects/moose_git/python/TestHarness/TestHarness.py", line 197, in findAndRunTests
    self.runner.run(tester, command)
  File "/Users/petejw/projects/moose_git/python/TestHarness/RunParallel.py", line 69, in run
    self.spinwait()
  File "/Users/petejw/projects/moose_git/python/TestHarness/RunParallel.py", line 159, in spinwait
    self.returnToTestHarness(job_index)
  File "/Users/petejw/projects/moose_git/python/TestHarness/RunParallel.py", line 135, in returnToTestHarness
    if not self.harness.testOutputAndFinish(tester, p.returncode, output, time, clock()):
  File "/Users/petejw/projects/moose_git/python/TestHarness/TestHarness.py", line 366, in testOutputAndFinish
    self.handleTestResult(tester.specs, output, result, start, end)
  File "/Users/petejw/projects/moose_git/python/TestHarness/TestHarness.py", line 546, in handleTestResult
    output = output.replace('\r', '\n')  # replace the carriage returns with newlines
AttributeError: 'bool' object has no attribute 'replace'
```

To continue developing on this branch: add a remote for me, fetch the branch, and continue making commits.

```
git remote add jwpeterson git@github.com:jwpeterson/moose.git
git fetch jwpeterson/amccaskey_5663
git checkout -b amccaskey_5663 jwpeterson/amccaskey_5663
```

We can then close this PR and open a new one on amccaskey/amccaskey_5663 when you are ready.  You should not run "git pull" for any reason while checked out on this branch, nor should you merge devel into this branch (or vide-versa) before making the PR.  We sometimes ask people to _rebase_ their work on the latest devel if it's old, but that should not be required here.

### CODE REVIEW

**Comment by jwpeterson** [comment]
The latest failure is due to a missing curl.h dependency, which means that Ubuntu does not have libcurl-devel packages installed by default.  Code works fine on my Mac, however.

This means we'll need to add a configure test for curl at the libmesh level and guard the ice_updater code with ifdefs on systems which don't have it...

```
In file included from /home/moosetest/moose_build/client_1_/moose/framework/contrib/ice_updater/src/LibcurlUtils.cpp:32:0:
/home/moosetest/moose_build/client_1_/moose/framework/contrib/ice_updater/include/LibcurlUtils.h:37:23: fatal error: curl/curl.h: No such file or directory
 #include <curl/curl.h>
```

File: framework/build.mk [hunk 1]
```diff
@@ -57,6 +57,7 @@ ifneq (,$(findstring mpi,$(cxx_compiler)))
 	cxx_compiler = $(shell $(libmesh_CXX) -show)
 endif
 
+
```
Comment (dschwen) [inline framework/build.mk]: This seems unnecessary

File: framework/src/base/Moose.C [hunk 2]
```diff
@@ -578,7 +579,6 @@ registerObjects(Factory & factory)
   registerPostprocessor(NodalExtremeValue);
   registerPostprocessor(ElementExtremeValue);
   registerPostprocessor(DifferencePostprocessor);
-  registerPostprocessor(FunctionValuePostprocessor);
```
Comment (dschwen) [inline framework/src/base/Moose.C]: What happened here? The include was just moved around, but the register call disappeared.

File: framework/include/outputs/ICEUpdater.h [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+  /**
+   * The Constructor
+   */
+  ICEUpdater(const InputParameters & params);
+
+  virtual ~ICEUpdater();
+
+  /**
+   * Performs basic error checking and initial setup of ExodusII_IO output object
```
Comment (dschwen) [inline framework/include/outputs/ICEUpdater.h]: That comment is not quite up-to-date

File: framework/build.mk (hunk 1 above)
Comment (jwpeterson) [inline framework/build.mk]: Thanks @dschwen. There are also a few unnecessary changes to Moose.C in these commits, probably due to merge conflict resolutions, which we'll get cleaned up before merging.

File: framework/src/base/Moose.C [hunk 4]
```diff
@@ -197,6 +196,7 @@
 #include "ExecutionerAttributeReporter.h"
 #include "PercentChangePostprocessor.h"
 #include "RealParameterReporter.h"
+#include "FunctionValuePostprocessor.h"
```
Comment (jwpeterson) [inline framework/src/base/Moose.C]: Moving this header file around is another spurious diff.

File: framework/src/base/Moose.C (hunk 2 above)
Comment (jwpeterson) [inline framework/src/base/Moose.C]: And we shouldn't delete this line.

**Comment by amccaskey** [comment]
@jwpeterson I fixed the issue with the python tester, the exodus mesh it was using never got committed. I changed the test to just use GeneratedMesh. I worked on the jwpeterson/amccaskey_5663 branch then committed the fixes to amccaskey/amccaskey_5663.

**Comment by jwpeterson** [comment]
> I worked on the jwpeterson/amccaskey_5663 branch then committed the fixes to amccaskey/amccaskey_5663.

OK, I cherry-picked your most recent three commits and tried them on OSX.  The test harness itself did not crash this time, but the test itself did not pass for me.  The error message is below:
```
outputs/iceupdater.testupdater............... FAILED (ICEUpdater FAILED: DID NOT GET CORRECT NUMBER OF POSTS)
Number of Posts was 0, but should have been 2
```
I am assuming this test does pass on your system, though...

**Comment by amccaskey** [comment]
Yea it's working on my OSX machine. I just pushed another update, try this one out if you can. You should see more debug info about the actual http post. Also I set it up so you can change the port in the tests input file in tests/outputs/iceupdater

**Comment by jwpeterson** [comment]
I tried your e3b9a8a00e, and get basically the same error:
```
outputs/iceupdater.testupdater............... FAILED (ICEUpdater FAILED: DID NOT GET CORRECT NUMBER OF POSTS)
Number of Posts was 0, but should have been 4
```

BTW, did you also add back mug.e for some reason in this same commit?

**Comment by jwpeterson** [comment]
One other thing: I'm not sure if your IDE is adding these "Signed-off-by:" lines to the commit messages, but we don't require them in MOOSE.

**Comment by amccaskey** [comment]
Yea Eclipse adds it - we have to keep that in for our ICE commits for IP reasons and I didn't think to switch it off. 

I did add that mesh back to try a more physically realistic problem, thought was maybe the execution was exiting too quickly to post anything meaningful to the server. 

Did you try changing the port in the tests and input file?

**Comment by jwpeterson** [comment]
> Did you try changing the port in the tests and input file?

Nope... what should I change it to?

**Comment by amccaskey** [comment]
I've been running on 8080 or 8181, you could try something like 56000 or something high that no other service is using. You can change it in test/tests/outputs/iceupdater/tests and test/tests/outputs/iceupdater/ice_updater_test.i (in the ICEUpdater block, url parameter). 

Also see if you see any error logs produced in test/tests/outputs/iceupdater/

**Comment by jwpeterson** [comment]
> Also see if you see any error logs produced in test/tests/outputs/iceupdater/

I just get the following:

[Mon Sep 21 12:26:37 2015] A POST request to http://localhost:8080/ice/update returned HTTP code 400.
[Mon Sep 21 12:26:37 2015] A POST request to http://localhost:8080/ice/update returned HTTP code 400.

Does this mean that SimpleHTTPServer isn't working for me?  Am I supposed to start up some kind server on my machine to get this test to pass?

**Comment by jwpeterson** [comment]
FYI, we get the same error on our CI box, which is now running because I ifdef'd out the Curl stuff and replaced it with error messages.  It did not seem to be used for anything, is that correct?
```
outputs/iceupdater.testupdater............... FAILED (ICEUpdater FAILED: DID NOT GET CORRECT NUMBER OF POSTS)
```

**Comment by dschwen** [comment]
If curl is not used that might explain why no messages are posted ;-)

**Comment by jwpeterson** [comment]
> If curl is not used that might explain why no messages are posted ;-)

Nice try, @dschewn.  The test also fails for me on my Mac, which has curl.h.  I actually put a mooseError in the constructor, so it should throw an error if it's even created, and it doesn't.

**Comment by amccaskey** [comment]
So the test fail on the CI makes sense to me since curl isn't available. The ICEUpdater test is going to try to execute a post, which tries to create a LibcurlUtils object, which will fail due to the mooseError you put in - nothing is getting posted to the python http server. 

As for your machine @jwpeterson I am confused as to why you're getting 400 error codes. The ICEUpdaterTester.py in TestHarness/testers starts up a http server when it is created in a separate thread, so you don't have to and you don't need ICE for the test. It's just using BaseHTTPServer.

I'll keep looking into it.

**Comment by amccaskey** [comment]
Let's give it one more try on your Mac. I've updated the server to send a code 200 response for a successful post.

**Comment by jwpeterson** [comment]
> Let's give it one more try on your Mac. I've updated the server to send a code 200 response for a successful post.

OK, I'm working on something else for a minute but I'll get back to you.

**Comment by jwpeterson** [comment]
With libcurl enabled, I get the following errors in the  updatererrors.log file with your latest commit:
```
[Mon Sep 21 16:32:08 2015] A POST request to http://localhost:8181/ice/update returned HTTP code 301.
[Mon Sep 21 16:32:08 2015] A POST request to http://localhost:8181/ice/update returned HTTP code 301.
[Mon Sep 21 16:32:09 2015] A POST request to http://localhost:8181/ice/update returned HTTP code 301.
[Mon Sep 21 16:32:10 2015] A POST request to http://localhost:8181/ice/update returned HTTP code 301.
```

I grabbed your latest commit, and I think I figured out the problem.  I have the environment variable `http_proxy` set in my environment -- we sit behind a proxy here , so we have to use this environment variable (and https_proxy, ftp_proxy, etc.) in order to use curl from the command line, but for some reason, it doesn't seem to work with `localhost`

If I `unset` that environment variable, the test now runs (and prints more stuff):
```
{'post': ['{"item_id":"1", "client_key":"1234567890ABCDEFGHIJ1234567890ABCDEFGHIJ", "posts":[{"type":"UPDATER_STARTED","message":""}]}']}
127.0.0.1 - - [21/Sep/2015 16:34:37] "POST /ice/update HTTP/1.1" 200 -
{'post': ['{"item_id":"1", "client_key":"1234567890ABCDEFGHIJ1234567890ABCDEFGHIJ", "posts":[{"type":"MESSAGE_POSTED","message":"NodalVariableValue10:0:0"}]}']}
127.0.0.1 - - [21/Sep/2015 16:34:37] "POST /ice/update HTTP/1.1" 200 -
{'post': ['{"item_id":"1", "client_key":"1234567890ABCDEFGHIJ1234567890ABCDEFGHIJ", "posts":[{"type":"MESSAGE_POSTED","message":"NodalVariableValue10:1:0.879017"}]}']}
127.0.0.1 - - [21/Sep/2015 16:34:38] "POST /ice/update HTTP/1.1" 200 -
{'post': ['{"item_id":"1", "client_key":"1234567890ABCDEFGHIJ1234567890ABCDEFGHIJ", "posts":[{"type":"MESSAGE_POSTED","message":"NodalVariableValue10:2:0.964154"},{"type":"UPDATER_STOPPED","message":""}]}']}
127.0.0.1 - - [21/Sep/2015 16:34:39] "POST /ice/update HTTP/1.1" 200 -
outputs/iceupdater.testupdater............................................................................ OK
```

**Comment by amccaskey** [comment]
Ok great, I'm glad you got it working. This is the correct output. 

Is this environment variable something we should worry about? Is it something that all MOOSE users have set, or just specific to your machine? 

What are the next steps for setting up Libmesh with configure test? Should we install curl on the CI machine?

**Comment by jwpeterson** [comment]
> Is this environment variable something we should worry about? Is it something that all MOOSE users have set, or just specific to your machine?

It's something that everyone at INL uses, so yeah it is something that we would have to work around.  Newer versions of curl support `curl --noproxy` so if there's a programmatic way to set this option through the API, that might be sufficient for us.

> What are the next steps for setting up Libmesh with configure test? 

This is pretty easy, I will work on it tomorrow, and this PR branch will then also include a libmesh update to that version.

> Should we install curl on the CI machine? 

Probably on at least some of them, so that this code gets tested.  I will also make the behavior of `LibcurlUtils` a bit more strict when curl is not found -- it should throw a `mooseError` when there is no curl, rather than just returning a string.

Also, would it be possible to license the contributed code under LGPL2.1?  It would make things simpler as I do not know if the "prior written permission" clause is incompatible with the license the rest of MOOSE is under...

```
3. Neither the name of the copyright holder nor the names of its 
   contributors may be used to endorse or promote products derived 
   from this software without specific prior written permission.
```

**Comment by jayjaybillings** [comment]
We have a whole slew of reasons why we don't release our code under LGPL or GPL and I'm reluctant to do it.

The Modified (3-clause) BSD license under which it is currently released (contributed) and the LGPL2.1 are completely compatible with each other and there aren't any problems releasing them together. It is a permissive license and works with just about all others. That is why I picked it for this work actually.

https://en.wikipedia.org/wiki/License_compatibility#GPL_compatibility

I think it must be OK for MOOSE too since 3/5 of the packages in framework/contribs are released under it. 

https://github.com/idaholab/moose/blob/devel/framework/contrib/README

**Comment by jayjaybillings** [comment]
And, for that matter, said permission granted. :)

**Comment by amccaskey** [comment]
I added the noproxy argument to LibcurlUtils setup for the POST call, let me know if that gets around the environment var problem you were having.

**Comment by dschwen** [comment]
I don't think the no proxy thing should be hard coded. There are potential additional uses for this, such as a website to show the simulation status where the user may want the outputer to respect the proxy settings. It is ready to make this a run time option.

**Comment by dschwen** [comment]
Also, do we need comments like "The constructor" and "The destructor" in our code? This was probably copy pasted from some MOOSE code but I think we should get rid of it.

**Comment by jayjaybillings** [comment]
I agree with Daniel that the proxy option shouldn't be hardcoded. I think
it should be something that ICE sets as an option in the input file or...
something. Just not hardcoded.

As far as the documentation goes, comments like that are the price of
working with students. We always put them so that the students can find
their way. I don't have a problem removing them if it fits the MOOSE style
better.

**Comment by amccaskey** [comment]
I too agree that it shouldn't be hardcoded, and did not intend to for the PR. With this last commit I'm just trying to help John diagnose what's going on with his machine and figure out future possible work-arounds (like this runtime option).

**Comment by jwpeterson** [comment]
I changed the `CURLOPT_NOPROXY` call to pass `"*"` for the third argument (this means don't use proxy for *any* host) and it now works on my machine even with `$http_proxy` set in the environment.  I also added the curl configure test to libmesh and updated MOOSE's libmesh submodule to this hash, so we can at least detect that curl is not available on the build machines.

The CI should, however, still fail on c58a98b since the ice updater test still runs even when `!LIBMESH_HAVE_CURL` (see remaining issues).

Remaining issues:

1. The noproxy thing should be made into a runtime option.  
1. I need to figure out why I don't see a `mooseError` from LibcurlUtils even when `!LIBMESH_HAVE_CURL`.  The test instead prints "FAILED (ICEUpdater FAILED: DID NOT GET CORRECT NUMBER OF POSTS)" when I would expect it to print my `mooseError`, "You tried to use the LibcurlUtils class, but CURL is not available."
1. Restrict the ice updater test in the test harness to only run when `LIBMESH_HAVE_CURL` is defined.  
1. When the ice updater test passes, it prints a bunch of output to the screen that we don't want -- the default for the test harness is to suppress all output for successful runs.

File: framework/contrib/ice_updater/include/LibcurlUtils.h [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+*******************************************************************************/
+#ifndef LIBCURLUTILS_H
+#define LIBCURLUTILS_H
+
+#include <iostream>
+#include <string>
+
+#include "libmesh/libmesh_config.h"
+
+#ifdef LIBMESH_HAVE_CURL
```
Comment (friedmud) [inline framework/contrib/ice_updater/include/LibcurlUtils.h]: Would it be better to just remove this entire class if Curl is not available?

If Curl is a requirement for this whole system - then all of these classes should just not exist if Curl is not available.

To get down to it... I don't really like the requirement of having curl.  Can we not find some lightweight open source C/C++ code that allows us to post/get http that we can just put in contrib directly?  I'm just worried that there are probably a lot of clusters out there without the Curl header files on them.

File: framework/contrib/ice_updater/include/LibcurlUtils.h [hunk 6]
```diff
@@ -0,0 +1,134 @@
+/*******************************************************************************
```
Comment (friedmud) [inline framework/contrib/ice_updater/include/LibcurlUtils.h]: I wonder if this stuff, that's actually agnostic of ICE, should go in its own `contrib` directory... we could use it in other places...

File: framework/contrib/ice_updater/include/Post.h [hunk 7]
```diff
@@ -0,0 +1,91 @@
+/*******************************************************************************
```
Comment (friedmud) [inline framework/contrib/ice_updater/include/Post.h]: Same with this one... we might use it in other instances... it's not ICE specific...

Maybe something like `contrib/url_tools` or something?

File: framework/contrib/ice_updater/include/LibcurlUtils.h (hunk 6 above)
Comment (jayjaybillings) [inline framework/contrib/ice_updater/include/LibcurlUtils.h]: +1 on its own directory. You may definitely have other uses for it.

CURL is the lightest thing we could find, but I didn't do an exhaustive
search. If you have suggestions, let us know and we will take a look.

On Tue, Sep 22, 2015 at 3:09 PM, Derek Gaston <notifications@github.com>
wrote:

> In framework/contrib/ice_updater/include/LibcurlUtils.h
> <https://github.com/idaholab/moose/pull/5717#discussion_r40130157>:
>
> > @@ -0,0 +1,134 @@
> > +/*******************************************************************************
>
> I wonder if this stuff, that's actually agnostic of ICE, should go in its
> own contrib directory... we could use it in other places...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/5717/files#r40130157>.
>



-- 
Jay Jay Billings
Oak Ridge National Laboratory
Twitter Handle: @jayjaybillings

Comment (permcody) [inline framework/contrib/ice_updater/include/LibcurlUtils.h]: I haven't looked but can we redistribute Curl sources ourselves?

File: framework/contrib/ice_updater/include/LibcurlUtils.h (hunk 5 above)
Comment (jwpeterson) [inline framework/contrib/ice_updater/include/LibcurlUtils.h]: > Would it be better to just remove this entire class if Curl is not available?

Depends how it's used... I didn't look at it closely.  But I generally prefer not making users put the #ifdefs in their code...

File: framework/contrib/ice_updater/include/LibcurlUtils.h (hunk 6 above)
Comment (friedmud) [inline framework/contrib/ice_updater/include/LibcurlUtils.h]: I suspect Curl is "hard" to build (meaning it can't just be put in `contrib`) (probably has it's own `configure` that really does need to be run).

Now that I look at the rest of this stuff... I actually wouldn't mind all of the ICE agnostic stuff just being directly in `framework/include/utils` or maybe `framework/include/utils/url` or something.  It already uses a lot of MOOSE stuff and it's weird that we now have stuff in `contrib` that is #including MOOSE stuff anyway.  It should really probably just be part of the framework.

That said, if we do that the license (at least on these parts) will definitely need to change to LGPL 2.1...

Comment (jwpeterson) [inline framework/contrib/ice_updater/include/LibcurlUtils.h]: > I haven't looked but can we redistribute Curl sources ourselves?

I don't think we want to mess with this.  There are different ways of building it (OpenSSL vs. TLS) I just don't want to deal with.  It is released under the MIT/X license, which means it can be used in commercial products.

File: framework/contrib/ice_updater/include/Updater.h [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+ * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+*******************************************************************************/
+#ifndef UPDATER_H
+#define UPDATER_H
+
+#include <string>
+#include <queue>
+#include <map>
+#include <tbb/mutex.h>
```
Comment (friedmud) [inline framework/contrib/ice_updater/include/Updater.h]: You can't rely on having TBB.  You can use `LIBMESH_HAVE_TBB_API`... but a better idea is to use (and add to if necessary) the abstractions in libMesh's  threads.h: https://github.com/libMesh/libmesh/blob/master/include/parallel/threads.h

That has abstractions in it for most things you want to do with TBB... but it works with both Pthreads and OpenMP as well.  You may need to add a bit for new things that are not yet there...

File: framework/contrib/ice_updater/include/Updater.h [hunk 9]
```diff
@@ -0,0 +1,320 @@
+/*******************************************************************************
```
Comment (friedmud) [inline framework/contrib/ice_updater/include/Updater.h]: Is this ICE specific?

File: python/TestHarness/testers/ICEUpdaterTester.py [hunk 10]
```diff
@@ -0,0 +1,88 @@
+from RunApp import RunApp
+from util import runCommand
+import os
+from BaseHTTPServer import BaseHTTPRequestHandler, HTTPServer
+from SocketServer import ThreadingMixIn
+import threading, cgi
+
+class ICEUpdaterTester(RunApp):
```
Comment (friedmud) [inline python/TestHarness/testers/ICEUpdaterTester.py]: This is pretty cool!

File: test/tests/outputs/iceupdater/ice_updater_test.i [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+[Mesh]
+  type = FileMesh
+  file = mug.e
+#dim = 3
+#  nx = 10
+#  ny = 10
+[]
+[Outputs]
+  file_base = out
+  [./ICEUpdater]
```
Comment (friedmud) [inline test/tests/outputs/iceupdater/ice_updater_test.i]: I like the way the syntax looks!

**Comment by dschwen** [comment]
@amccaskey, where is the client key set? And what is the "property file"?

File: framework/contrib/ice_updater/include/ErrorLogger.h [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+  void logError(string error);
+
+  /**
+   * Dumps the errors in errorVector into a file with a filename format
+   * "updatererrors_&lt;timestamp&gt;.log".
+   *
+   * @return The name of the error log file.
+   */
+  string dumpErrors();
+
```
Comment (dschwen) [inline framework/contrib/ice_updater/include/ErrorLogger.h]: You have a ton of empty lines after ```{``` and before ```}``` which is a bit at odds with our [code standards(http://mooseframework.org/wiki/CodeStandards/). Same goes for single line doc-comments not using ```///```, using ```using std;```, not having a whitespace after ```//```, and indentation (switch/case for example).
But I know I _am_ being picky here ;-)

File: framework/contrib/ice_updater/include/ErrorLogger.h [hunk 13]
```diff
@@ -0,0 +1,84 @@
+/*******************************************************************************
```
Comment (dschwen) [inline framework/contrib/ice_updater/include/ErrorLogger.h]: So.. why do we have a custom error logger here? I noticed parts of the contrib stuff just call mooseError. This seems simpler (that or mooseWarning).

File: framework/contrib/ice_updater/include/ErrorLogger.h (hunk 12 above)
Comment (jwpeterson) [inline framework/contrib/ice_updater/include/ErrorLogger.h]: @dschwen this is in contrib, so we don't need to apply the normal coding conventions IMO.

Comment (friedmud) [inline framework/contrib/ice_updater/include/ErrorLogger.h]: I was thinking along the same lines... but this is all custom code for MOOSE... that might end up in MOOSE (I kind of think it should already be put in MOOSE).

It could be good to go ahead and enforce our style now (but that's easy for me to say sense I'm not doing the code changes!).

**Comment by amccaskey** [comment]
@dschwen in ICE you could conceivably communicate with a number of different client instances that interact with the running core infrastructure. The key is to indicate which one you want the Updater to communicate with. Now, in most cases we just interact with one client, so the client_key is hard-coded in ICEUpdater. In the future we could set this with another input file parameter. 

As for the properties file - we initially developed this for other applications that required the Updater be initialized with an input properties file. That file just contains all the key value pairs you see in ICEUpdater's constructor. For this work it was just easier to input them programmatically from the InputParameters than from a properties file.

File: framework/contrib/ice_updater/include/LibcurlUtils.h (hunk 6 above)
Comment (friedmud) [inline framework/contrib/ice_updater/include/LibcurlUtils.h]: How about the idea of putting all of the ICE agnostic stuff in `framework/include/utils/url` or something?  It basically is MOOSE code... there's no real reason to keep it separated over in the contrib area.  It wasn't an external project that we brought in or anything...

**Comment by friedmud** [comment]
@amccaskey might be best to make that client_key an input parameter with a default value...

**Comment by jwpeterson** [comment]
So 955d0b7 passed the normal MOOSE and parallel tests, but the TestHarness crashed during RECOVER testing.  I guess it might be due to the new kind of Tester that was added?  @permcody any ideas?

```
Traceback (most recent call last):
  File "./run_tests", line 24, in <module>
    TestHarness.buildAndRun(sys.argv, app_name, MOOSE_DIR)
  File "/home/moosetest/moose_build/client_0_/moose/python/TestHarness/TestHarness.py", line 34, in buildAndRun
    harness.findAndRunTests()
  File "/home/moosetest/moose_build/client_0_/moose/python/TestHarness/TestHarness.py", line 171, in findAndRunTests
    testers = self.appendRecoverableTests(testers)
  File "/home/moosetest/moose_build/client_0_/moose/python/TestHarness/TestHarness.py", line 310, in appendRecoverableTests
    part2 = copy.deepcopy(part1)
  File "/usr/lib/python2.6/copy.py", line 189, in deepcopy
    y = _reconstruct(x, rv, 1, memo)
  File "/usr/lib/python2.6/copy.py", line 338, in _reconstruct
    state = deepcopy(state, memo)
  File "/usr/lib/python2.6/copy.py", line 162, in deepcopy
    y = copier(x, memo)
  File "/usr/lib/python2.6/copy.py", line 255, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/usr/lib/python2.6/copy.py", line 162, in deepcopy
    y = copier(x, memo)
  File "/usr/lib/python2.6/copy.py", line 292, in _deepcopy_inst
    state = deepcopy(state, memo)
  File "/usr/lib/python2.6/copy.py", line 162, in deepcopy
    y = copier(x, memo)
  File "/usr/lib/python2.6/copy.py", line 255, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/usr/lib/python2.6/copy.py", line 189, in deepcopy
    y = _reconstruct(x, rv, 1, memo)
  File "/usr/lib/python2.6/copy.py", line 338, in _reconstruct
    state = deepcopy(state, memo)
  File "/usr/lib/python2.6/copy.py", line 162, in deepcopy
    y = copier(x, memo)
  File "/usr/lib/python2.6/copy.py", line 255, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/usr/lib/python2.6/copy.py", line 189, in deepcopy
    y = _reconstruct(x, rv, 1, memo)
  File "/usr/lib/python2.6/copy.py", line 338, in _reconstruct
    state = deepcopy(state, memo)
  File "/usr/lib/python2.6/copy.py", line 162, in deepcopy
    y = copier(x, memo)
  File "/usr/lib/python2.6/copy.py", line 255, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/usr/lib/python2.6/copy.py", line 188, in deepcopy
    "un(deep)copyable object of type %s" % cls)
copy.Error: un(deep)copyable object of type <type 'thread.lock'>
ERROR: Cleaning up
```

File: python/TestHarness/testers/ICEUpdaterTester.py [hunk 14]
```diff
@@ -0,0 +1,88 @@
+from RunApp import RunApp
+from util import runCommand
+import os
+from BaseHTTPServer import BaseHTTPRequestHandler, HTTPServer
+from SocketServer import ThreadingMixIn
```
Comment (permcody) [inline python/TestHarness/testers/ICEUpdaterTester.py]: OMG - Socket Threading!!!!

**Comment by permcody** [comment]
Gotta love Python... A Python type "thread.lock" actually crashes inside of
another python library dict.deepcopy...  Oh well, we just need to skip
recover for this new Tester.

**Comment by jwpeterson** [comment]
The moose_PR_app_tests failure in 55da16b is related to the libmesh update, not the ICEUpdater stuff.  @dschwen is looking into it.

**Comment by dschwen** [comment]
Yeah that is a test that needs to be re-golded because libMesh contains an automatic differentiation fix. This was expected and is mentioned here: https://github.com/libMesh/libmesh/pull/653

File: framework/contrib/ice_updater/include/LibcurlUtils.h (hunk 6 above)
Comment (jayjaybillings) [inline framework/contrib/ice_updater/include/LibcurlUtils.h]: Fine by me and if it goes into the framework, yes, the license must change
to LGPL. Totally cool since that would just be MOOSE stuff that we are
using and won't redistribute.

**Comment by jwpeterson** [comment]
I think I'm pretty much done with this PR...  In the future, we can think more about moving the code from contrib to MOOSE itself if that's desired, it now basically conforms to all the usual MOOSE coding conventions. The `Updater` code is now protected by `LIBMESH_HAVE_TBB_API` guards, but is not yet implemented in terms of more generic `Threads` interfaces -- it uses some fairly specialized aspects of TBB that I'm not sure we've abstracted yet.

**Comment by jayjaybillings** [comment]
Works for me. Alex would probably be happy to help with the migration of
the code to the framework.

Thanks for helping us get this in!

**Comment by friedmud** [comment]
Even if we don't move this to the framework right now... I would vote for
at least changing the license on the code we want to move... otherwise
we'll have to "relicense" it when we do move it (getting permission from
the copyright holder).

Honestly, if this is something we want in MOOSE (which I think we do) we
should really just move it now and avoid the hassle...

**Comment by dschwen** [comment]
Not to open another can of worms, but should this be called ```ICEUpdater``` or maybe rather something like ```HTTPPostOutput```? I for one plan on writing a web app to monitor my MOOSE runs with.

**Comment by jayjaybillings** [comment]
The parts that move into MOOSE proper can be called whatever you want.

On Thu, Sep 24, 2015 at 6:40 PM, Daniel Schwen <notifications@github.com>
wrote:

> Not to open another can of worms, but should this be called ICEUpdater or
> maybe rather something like HTTPPostOutput? I for one plan on writing a
> web app to monitor my MOOSE runs with.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/5717#issuecomment-143068220>.
>



-- 
Jay Jay Billings
Oak Ridge National Laboratory
Twitter Handle: @jayjaybillings

**Comment by friedmud** [comment]
@dschwen Most of the classes are already pretty generic... I suggested calling this collection of classes `url_tools` or something... and putting it in `utils`.

It's really is pretty nice code for doing JSON post/gets... which really will be handy over the years for all kinds of purposes.  Thanks @amccaskey !

**Comment by jayjaybillings** [comment]
Honestly, if we have to change the license immediately then this is going
to be delayed because I will have to review the whole code and figure out
what can, can't and might could be relicensed. I just don't have time to do
that now. Since I'm the PI for the copyright holder I will happily promise
to revisit it later. Probably a couple of months from now.

On Thu, Sep 24, 2015 at 6:33 PM, Derek Gaston <notifications@github.com>
wrote:

> Even if we don't move this to the framework right now... I would vote for
> at least changing the license on the code we want to move... otherwise
> we'll have to "relicense" it when we do move it (getting permission from
> the copyright holder).
>
> Honestly, if this is something we want in MOOSE (which I think we do) we
> should really just move it now and avoid the hassle...
>
> On Thu, Sep 24, 2015 at 4:37 PM Jay Jay Billings <notifications@github.com
> >
> wrote:
>
> > Works for me. Alex would probably be happy to help with the migration of
> > the code to the framework.
> >
> > Thanks for helping us get this in!
> > On Sep 24, 2015 3:53 PM, "John W. Peterson" <notifications@github.com>
> > wrote:
> >
> > > I think I'm pretty much done with this PR... In the future, we can
> think
> > > more about moving the code from contrib to MOOSE itself if that's
> > desired,
> > > it now basically conforms to all the usual MOOSE coding conventions.
> The
> > > Updater code is now protected by LIBMESH_HAVE_TBB_API guards, but is
> not
> > > yet implemented in terms of more generic Threads interfaces -- it uses
> > > some fairly specialized aspects of TBB that I'm not sure we've
> abstracted
> > > yet.
> > >
> > > —
> > > Reply to this email directly or view it on GitHub
> > > <https://github.com/idaholab/moose/pull/5717#issuecomment-143031848>.
> > >
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/idaholab/moose/pull/5717#issuecomment-143044031>.
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/5717#issuecomment-143067323>.
>



-- 
Jay Jay Billings
Oak Ridge National Laboratory
Twitter Handle: @jayjaybillings

**Comment by friedmud** [comment]
Sounds good @jayjaybillings ... I think that closes the issue :-)

**Comment by jwpeterson** [comment]
> Not to open another can of worms, but should this be called ICEUpdater or maybe rather something like HTTPPostOutput? I for one plan on writing a web app to monitor my MOOSE runs with.

@dschwen Might be a good topic for another PR.
